pallets/external-validators/src/lib.rs

```diff
@@ -0,0 +1,697 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+//! ExternalValidators pallet.
+//!
+//! A pallet to manage external validators for a solochain.
+//!
+//! ## Terminology
+//!
+//! - WhitelistedValidators: Fixed validators set by root/governance. Have priority over the external validators.
+//! - ExternalValidators: Validators set using storage proofs from another blockchain. Changing them triggers a
+//!     new era. Can be disabled by setting `SkipExternalValidators` to true.
+//!
+//! Validators only change once per era. By default the era changes after a fixed number of sessions, but that can
+//! be changed by root/governance to increase era on every session, or to never change.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+use {
+    frame_support::pallet_prelude::Weight,
+    log::log,
+    parity_scale_codec::{Decode, Encode, MaxEncodedLen},
+    scale_info::TypeInfo,
+    sp_runtime::traits::Get,
+    sp_runtime::RuntimeDebug,
+    sp_staking::SessionIndex,
+    sp_std::collections::btree_set::BTreeSet,
+    sp_std::vec::Vec,
+    tp_traits::{
+        ActiveEraInfo, EraIndex, EraIndexProvider, InvulnerablesProvider, OnEraEnd, OnEraStart,
+        ValidatorProvider,
+    },
+};
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+
+#[frame_support::pallet]
+pub mod pallet {
+    pub use crate::weights::WeightInfo;
+
+    #[cfg(feature = "runtime-benchmarks")]
+    use frame_support::traits::Currency;
+
+    use {
+        super::*,
+        frame_support::{
+            dispatch::DispatchResultWithPostInfo,
+            pallet_prelude::*,
+            traits::{EnsureOrigin, UnixTime, ValidatorRegistration},
+            BoundedVec, DefaultNoBound,
+        },
+        frame_system::pallet_prelude::*,
+        sp_runtime::{traits::Convert, SaturatedConversion},
+        sp_std::vec::Vec,
+    };
+
+    /// The current storage version.
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+    /// A convertor from collators id. Since this pallet does not have stash/controller, this is
+    /// just identity.
+    pub struct IdentityCollator;
```

Comment:
this should not live here, nor with such naming

Reply:
*fixed*